extensions/mqtt/processors/PublishMQTT.cpp

```diff
@@ -34,76 +34,229 @@ void PublishMQTT::initialize() {
+  message_to_publish.payload = const_cast<std::byte*>(buffer.data());
+  message_to_publish.payloadlen = buffer.size();
+  message_to_publish.qos = qos_.value();
+  message_to_publish.retained = retain_;
+
+  setMqtt5Properties(message_to_publish, content_type, flow_file);
+
+  MQTTAsync_responseOptions response_options = MQTTAsync_responseOptions_initializer;
+  if (mqtt_version_ == MqttVersions::V_5_0) {
+    response_options.onSuccess5 = sendSuccess5;
+    response_options.onFailure5 = sendFailure5;
   } else {
-    logger_->log_debug("Sent flow with length %d to MQTT topic %s", callback.read_size_, topic_);
-    session->transfer(flowFile, Success);
-  }
-}
-
-int64_t PublishMQTT::ReadCallback::operator()(const std::shared_ptr<io::InputStream>& stream) {
-  if (flow_size_ < max_seg_size_)
-    max_seg_size_ = flow_size_;
-  gsl_Expects(max_seg_size_ < gsl::narrow<uint64_t>(std::numeric_limits<int>::max()));
-  std::vector<std::byte> buffer(max_seg_size_);
-  read_size_ = 0;
-  status_ = 0;
-  while (read_size_ < flow_size_) {
-    // MQTTClient_message::payloadlen is int, so we can't handle 2GB+
-    const auto readRet = stream->read(buffer);
-    if (io::isError(readRet)) {
-      status_ = -1;
-      return gsl::narrow<int64_t>(read_size_);
+    response_options.onSuccess = sendSuccess;
+    response_options.onFailure = sendFailure;
+  }
+
+  // save context for callback
+  std::packaged_task<bool(bool, std::optional<int>, std::optional<MQTTReasonCodes>)> send_finished_task(
+          [this] (const bool success, const std::optional<int> response_code, const std::optional<MQTTReasonCodes> reason_code) {
+            return notify(success, response_code, reason_code);
+          });
+  response_options.context = &send_finished_task;
+
+  in_flight_message_counter_.increase();
+
+  const int error_code = MQTTAsync_sendMessage(client_, topic.c_str(), &message_to_publish, &response_options);
+  if (error_code != MQTTASYNC_SUCCESS) {
+    logger_->log_error("MQTTAsync_sendMessage failed on topic '%s', MQTT broker %s with error code [%d]", topic, uri_, error_code);
+    // early fail, sending attempt did not succeed, no need to wait for callback
+    in_flight_message_counter_.decrease();
+    return false;
+  }
+
+  return send_finished_task.get_future().get();
+}
+
+void PublishMQTT::checkProperties() {
+  if ((mqtt_version_ == MqttVersions::V_3_1_0 || mqtt_version_ == MqttVersions::V_3_1_1 || mqtt_version_ == MqttVersions::V_3X_AUTO)) {
+    if (isPropertyExplicitlySet(MessageExpiryInterval)) {
+      logger_->log_warn("MQTT 3.x specification does not support Message Expiry Intervals. Property is not used.");
+    }
+    if (isPropertyExplicitlySet(ContentType)) {
+      logger_->log_warn("MQTT 3.x specification does not support Content Types. Property is not used.");
+    }
+  }
+}
+
+void PublishMQTT::checkBrokerLimitsImpl() {
+  if (retain_available_.has_value() && !*retain_available_ && retain_) {
```

Comment:
I would do
```suggestion
  if (retain_available_ == false && retain_) {
```
here, too

Reply:
Done.